### PR TITLE
Add Daml library function support for hex string

### PR DESCRIPTION
**Pull request by carlpulley-da**
- [x] Add in `pack` and `slice` hex string functions
- [x] Add in validation functions for `Bytes32`, `UInt32`, `UInt64` and `UInt256`
- [x] Add in hex encodings/decodings for `Numeric n`, `Party` and `Int64`
- [ ] ~~Add in hex encodings/decodings for `ContractId a`~~
    - Created https://github.com/DACH-NY/canton-network-utilities/issues/3852 (`BETextToAnyContractId: BytesHex -> AnyContractId` primitive) to track this issue
- [ ] ~~Add in Ethereum address validation function~~
    - Based on https://eips.ethereum.org/EIPS/eip-55, ethereum address *strings* can be mixed case hex strings (which encodes a keccak256 based hash within the string). For the current PR, we likely do not want to implement such a function.
    - Created https://github.com/DACH-NY/canton-network-utilities/issues/3793 (`isValidEthereumAddress`) to track this
- [x] Add in testing

### CODE REVIEW

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 1]
```diff
@@ -78,14 +87,106 @@ class HasToHex a where
...
+  toHex 14 = "0e"
+  toHex 15 = "0f"
+  toHex n | n < 256 = intToHexByte n
+  toHex n = (toHex(n / 256)) <> (toHex(n % 256))
+
+instance HasToHex Party where
+  toHex p = toHex(partyToText p)
+
+instance NumericScale n => HasToHex (Numeric n) where
+  toHex num = toHex(show num)
```
Comment (meiersi-da) [inline sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:120]: Isn't this a rather surprising instance? I'd suggest not offering that one. It seems very confusing to hexEncode the utf-8 encoding of the decimal string representation.

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 2]
```diff
@@ -78,14 +87,106 @@ class HasToHex a where
...
+
 class HasFromHex a where
   -- | Converts a hex encoded string into a typed data value.
   fromHex: BytesHex -> a
 
-instance HasToHex Text where
-  toHex = primitive @"BEEncodeHex"
-
 instance HasFromHex Text where
   fromHex "" = ""
```
Comment (meiersi-da) [inline sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:127]: Why is this required?

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 3]
```diff
@@ -78,14 +87,106 @@ class HasToHex a where
...
+  fromHex hex | Text.isHex hex =
+    let right = fromHex(Text.substring ((Text.length hex) - 2) 2 hex)
+        left = fromHex(Text.substring 0 ((Text.length hex) - 2) hex)
+    in
+      left * 256 + right
+  fromHex arg = error $ "Expected a hex encoded string but argument was: " <> arg
+
+instance HasFromHex Party where
+  fromHex hex | Text.isHex hex = fromSome(partyFromText(fromHex hex))
+  fromHex arg = error $ "Expected a hex encoded string but argument was: " <> arg
```
Comment (meiersi-da) [inline sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:143]: I'm not convinced about pretending that `fromHex` is a total function when it really is not. Now as long as the Daml code doesn't need to recover from the error this is fine.

In particular, if all the primitives used are exposed, so that users can write their own parsers.

So fine for a 3.3 release.

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 4]
```diff
@@ -78,14 +87,106 @@ class HasToHex a where
...
+instance HasToHex Int where
+  toHex n | 0 <= n && n <= 9 = "0" <> (show n)
+  toHex 10 = "0a"
+  toHex 11 = "0b"
+  toHex 12 = "0c"
+  toHex 13 = "0d"
+  toHex 14 = "0e"
+  toHex 15 = "0f"
+  toHex n | n < 256 = intToHexByte n
+  toHex n = (toHex(n / 256)) <> (toHex(n % 256))
```
Comment (meiersi-da) [inline sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:114]: This repeated concatentation has quadratic cost in the lenght of the output. Apart from building a primitive, the better option is to build the list from right to left, and then call `Text.implode` (https://docs.daml.com/daml/stdlib/DA-Text.html#function-da-text-implode-82253)

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 5]
```diff
@@ -78,14 +87,106 @@ class HasToHex a where
...
+
+-- | Validate that the byte encoded string is UInt256Hex
+isUInt256Hex: BytesHex -> Bool
+isUInt256Hex hex | Text.isHex hex = byteCount hex == 8
+isUInt256Hex arg = error $ "Expected a hex encoded string but argument was: " <> arg
+
+-- | HIDE
+constList: Int -> a -> [a]
+constList 0 _ = []
+constList n x = [x] ++ (constList (n - 1) x)
```
Comment (meiersi-da) [inline sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:178]:
```suggestion
constList n x = x :: (constList (n - 1) x)
```

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    sliceHexBytes "DeadBeef" 2 5 === "adBeef"
+    sliceHexBytes "DeadBeef" 3 5 === "Beef"
+
+    -- isBytes32Hex, isUInt32Hex, isUInt64Hex, isUInt256Hex
+    isBytes32Hex(packHexBytes "00" 32) === True
+    isBytes32Hex(packHexBytes "DeadBeef" 32) === True
+    isUInt32Hex(packHexBytes "00" 5) === True
+    isUInt32Hex(packHexBytes "000001" 5) === True
+    isUInt64Hex(packHexBytes "00" 6) === True
+    isUInt256Hex(packHexBytes "00" 8) === True
```
Comment (meiersi-da) [inline sdk/compiler/damlc/tests/daml-test-files/HexString.daml:93]: negative tests?

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    packHexBytes "01" 3 === "000001"
+    packHexBytes "0102" 3 === "000102"
+
+    -- sliceHexBytes
+    sliceHexBytes "12" 1 2 === "12"
+    sliceHexBytes "DeadBeef" 2 3 === "ad"
+    sliceHexBytes "DeadBeef" 3 4 === "Be"
+    sliceHexBytes "DeadBeef" 2 4 === "adBe"
+    sliceHexBytes "DeadBeef" 2 5 === "adBeef"
+    sliceHexBytes "DeadBeef" 3 5 === "Beef"
```
Comment (meiersi-da) [inline sdk/compiler/damlc/tests/daml-test-files/HexString.daml:85]: out of bounds tests?

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    toHex "" === ""
+    toHex " " === "20"
+    toHex "a" === "61"
+    toHex "Hello World!" === "48656c6c6f20576f726c6421"
+    toHex "DeadBeef" === "4465616442656566"
+    fromHex "" === ""
+    fromHex "20" === " "
+    fromHex "61" === "a"
+    fromHex "48656c6c6f20576f726c6421" === "Hello World!"
+    fromHex "4465616442656566" === "DeadBeef"
```
Comment (meiersi-da) [inline sdk/compiler/damlc/tests/daml-test-files/HexString.daml:53]: max and min bounds for the different int types.

**Review by meiersi-da — APPROVED** [review]
Thx. I expect we'll want to make another pass over these functions as we use them in anger. However, for a 3.3 alpha feature this seems good enough.

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml (hunk 2 above)
Comment (carlpulley-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:127]: I'd been thinking about the `fromHex(toHex "") =?= ""` equation when I added this case.

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml (hunk 3 above)
Comment (carlpulley-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:143]: I'll refactor so that the `HasFromHex a` instances are defined for `HasFromHex (Optional a)` instances instead

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml (hunk 1 above)
Comment (carlpulley-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:120]: Agreed it does look odd.

However, the numeric is (essentially) a string represented decimal (with use sites knowing what scale it should be interpreted as).

By exploiting this decimal string representation, we simplify the corresponding `fromHex` instance be an application of `parseNumeric`.

Perhaps a comment here would help to better understand this rationale?

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml (hunk 8 above)
Comment (carlpulley-da) [reply sdk/compiler/damlc/tests/daml-test-files/HexString.daml]: Could you expand on this please? I presume you mean for the `isBytes32Hex`, `isUInt32Hex`, etc. range of functions?

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml (hunk 1 above)
Comment (meiersi-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:130]: The rule I tend to use is to not define instances that are surprising / for which there are multiple options. For these, I instead define aptly named functions. For example,
```
-- | Convert a decimal to a hex-bytes representation by rendering it as a decimal string
-- and encoding that one using UTF-8.
decimalViaStringToHex : Decimal -> Hex
decimalVaiStringFromHex : Hex -> Optional Decimal
```

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml (hunk 8 above)
Comment (meiersi-da) [reply sdk/compiler/damlc/tests/daml-test-files/HexString.daml]: yes

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 9]
```diff
@@ -78,14 +96,157 @@ class HasToHex a where
...
+constList: Int -> a -> [a]
+constList 0 _ = []
+constList n x = x :: (constList (n - 1) x)
+
+-- | HIDE
+byteRepeat: Int -> BytesHex -> BytesHex
+byteRepeat len byte | Text.isHex byte && byteCount byte == 1 && len >= 0 = Text.implode(constList len byte)
+byteRepeat _ arg | not (Text.isHex arg) = error $ "Expected a hex encoded string but argument was: " <> arg
+byteRepeat _ hex | Text.isHex hex && byteCount hex /= 1 = error $ "Expected a byte value but hex encoded string was: " <> hex
+byteRepeat len _ = error $ "Expected a positive length but argument was: " <> show len
```
Comment (remyhaemmerle-da) [inline sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:190]: I found the difficult to parse. 
I would rather use standard Haskell way

I would rather put the failing case first, and use the standard `otherwise` guard for the validated case. 
```suggestion
byteRepeat: Int -> BytesHex -> BytesHex
byteRepeat len arg 
  | not (Text.isHex arg) = error $ "Expected a hex encoded string but argument was: " <> arg
  |  byteCount hex /= 1  = error $ "Expected a byte value but hex encoded string was: " <> hex
  | len < 0                      = error $ "Expected a positive length but argument was: " <> show len
  | otherwise                 = Text.implode(constList len byte)
```

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+  script do
+    -- `HasToHex Int` and `HasFromHex Int`
+    toHex 0 === "00"
+    toHex 1 === "01"
+    toHex 15 === "0f"
+    toHex 16 === "10"
+    toHex 255 === "ff"
+    toHex 256 === "0100"
+    toHex 257 === "0101"
+    toHex 3735928559 === "deadbeef"
```
Comment (remyhaemmerle-da) [inline sdk/compiler/damlc/tests/daml-test-files/HexString.daml:32]: I am missing test for 
`-1`, `maxBound @Int`, and `minBound @Int`

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    fromHex "" === Some 0
+    fromHex "00" === Some 0
+    fromHex "01" === Some 1
+    fromHex "0a" === Some 10
+    fromHex "0F" === Some 15
+    fromHex "10" === Some 16
+    fromHex "Ff" === Some 255
+    fromHex "0100" === Some 256
+    fromHex "0101" === Some 257
+    fromHex "DeadBeef" === Some 3735928559
```
Comment (remyhaemmerle-da) [inline sdk/compiler/damlc/tests/daml-test-files/HexString.daml:42]: we need test for all the exceptions cases.

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    alice <- allocateParty "Alice"
+    Some alice === fromHex(toHex alice)
+
+    -- `numericViaStringToHex` and `numericViaStringFromHex`
+    numericViaStringToHex n3 === "302e313233"
+    numericViaStringToHex n7 === "302e31323334353637"
+    numericViaStringToHex n10 === "302e313233343536373839"
+    numericViaStringFromHex "302e313233" === Some n3
+    numericViaStringFromHex "302e31323334353637" === Some n7
+    numericViaStringFromHex "302e313233343536373839" === Some n10
```
Comment (remyhaemmerle-da) [inline sdk/compiler/damlc/tests/daml-test-files/HexString.daml:66]: we need test for,`(0 : Numeric n), (-1 : Numeric n), (1: Numeric n), (maxBound @(Numeric n)), (minBound @(Numeric n)) for n = 0 and n=37 at least.
cases where the Numeric under/overflow by 1 and by a lot.

File: sdk/docs/sharable/sdk/reference/daml/stdlib/Prelude.rst [hunk 13]
```diff
@@ -1523,6 +1523,10 @@ Data Types
 
   The ``Party`` type represents a party to a contract\.
 
+  **instance** :ref:`HasFromHex <class-da-crypto-text-hasfromhex-84972>` (`Optional <type-da-internal-prelude-optional-37153_>`_ `Party <type-da-internal-lf-party-57932_>`_)
+
+  **instance** :ref:`HasToHex <class-da-crypto-text-hastohex-92431>` `Party <type-da-internal-lf-party-57932_>`_
+
```
Comment (remyhaemmerle-da) [inline sdk/docs/sharable/sdk/reference/daml/stdlib/Prelude.rst:1529]: please mode the classes `HasFromHex` and `HasToHex` from the `DA.Text` to `DA.Crypto.Text`

**Review by remyhaemmerle-da — APPROVED** [review]
functions from the standard library should be extensively test.
in particular we should test all the corner cases, and all the exception cases.
This can be done in upcoming PR.

File: sdk/docs/sharable/sdk/reference/daml/stdlib/Prelude.rst (hunk 13 above)
Comment (remyhaemmerle-da) [reply sdk/docs/sharable/sdk/reference/daml/stdlib/Prelude.rst:1529]: in fact move also `DA.isHex`. 
Since the feature is alpha, we may want to delete/change anything we defined for cctp. it would be easier if those function are not define in stable package.

**Comment by remyhaemmerle-da** [comment]
> However, for a 3.3 alpha feature this seems good enough.

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 14]
```diff
@@ -78,14 +96,157 @@ class HasToHex a where
...
+int16ToHexByte 10 = "a"
+int16ToHexByte 11 = "b"
+int16ToHexByte 12 = "c"
+int16ToHexByte 13 = "d"
+int16ToHexByte 14 = "e"
+int16ToHexByte 15 = "f"
+int16ToHexByte n = error $ "Int16 should be in range [0, 16): " <> show n
+
+-- | HIDE
+int256ToHexByte: Int -> BytesHex
```
[the remark this answers is not in the document]
Comment (carlpulley-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:114]: Renamed to `uint256ToHexByte` as this is consistent with the `uint16ToHexByte` function name. In this instance, I'd prefer not to name it `byteToHexByte` as I feel this is confusing (too many bytes indirections in the name).

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 15]
```diff
@@ -78,14 +96,157 @@ class HasToHex a where
...
+-- | Minimum UInt32 hex value
+minUInt32Hex: BytesHex
+minUInt32Hex = byteRepeat 5 "00"
+-- | Maximum UInt32 hex value
+maxUInt32Hex: BytesHex
+maxUInt32Hex = byteRepeat 5 "ff"
+
+-- | Validate that the byte encoded string is UInt32Hex
+isUInt32Hex: BytesHex -> Bool
+isUInt32Hex hex | Text.isHex hex = byteCount hex == 5
```
Comment (dylant-da) [inline sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:213]: I'm not sure if I'm misunderstanding the logic here - shouldn't Uint32 require only 4 bytes to encode, Uint64 require 8 bytes, and Uint256 require 32?

i.e.
```
ghci> logBase 256 (2 ** 32)
4.0
ghci> logBase 256 (2 ** 64)
8.0
ghci> logBase 256 (2 ** 256)
32.0
```

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+    sliceHexBytes "DeadBeef" 0 2 === Left "Expected start byte to be >= 1, was: 0"
+    sliceHexBytes "DeadBeef" 3 6 === Left "Expected end byte to be >= 3 and <= 5 was: 6"
+    sliceHexBytes "DeadBeef" 3 2 === Left "Expected end byte to be >= 3 and <= 5 was: 2"
+
+    -- isBytes32Hex, isUInt32Hex, isUInt64Hex, isUInt256Hex
+    isBytes32Hex(fromSome(packHexBytes "00" 32)) === True
+    isBytes32Hex(fromSome(packHexBytes "DeadBeef" 32)) === True
+    isBytes32Hex(fromSome(packHexBytes "00" 31)) === False
+    isBytes32Hex(fromSome(packHexBytes "00" 33)) === False
+    isBytes32Hex(minBytes32Hex) === True
```
Comment (dylant-da) [inline sdk/compiler/damlc/tests/daml-test-files/HexString.daml:96]: You can drop the parens here, e.g.
```suggestion
    isBytes32Hex minBytes32Hex === True
```

**Review by dylant-da — APPROVED** [review]
Thanks for the hard work on this! The main worry for me in this review is my comment on the widths of Uint32/64/256 - please make sure to address that before merging.

I also have some parenthesization nitpicks - if I were to annotate every place where unnecessary parens surround arguments there'd be too many comments - I've put comments at the first few examples, I leave it up to you to extrapolate to the rest of the code.

Feel free to ignore the stylistic choices on guards if you feel they're irrelevant.

Otherwise, LGTM!

**Comment by remyhaemmerle-da** [comment]
We agreed offline on tackling only the following changes in the PR:
- Move the definitions of `HasToHex` and `HasFromHex` classes from DA.Text to DA.Crypto.Text.
- Add a warning when loading `DA.Crypto.Text`

Other changes will be addressed after the 3.3 cutoff.

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml (hunk 14 above)
[the remark this answers is not in the document]
Comment (carlpulley-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:114]: bytes32 is 32 bytes

confusingly, uint16 and uint256 are bit based - i.e. 16 bit or 256 bit uint.

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml (hunk 15 above)
Comment (carlpulley-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:213]: Begger! Thanks for pointing that error out!

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 17]
```diff
@@ -78,14 +96,157 @@ class HasToHex a where
...
+numericViaStringFromHex _ = None
+
+-- | Number of bytes present in a byte encoded string.
+byteCount: BytesHex -> Int
+byteCount "" = 0
+byteCount hex | Text.isHex hex = (Text.length hex) / 2
+byteCount arg = error $ "Expected a hex encoded string but argument was: " <> arg
+
+-- | HIDE
+constList: Int -> a -> [a]
```
[the remark this answers is not in the document]
Comment (carlpulley-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:181]: Will refactor to use the name `replicateList`

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml [hunk 18]
```diff
@@ -78,14 +96,157 @@ class HasToHex a where
...
+int256ToHexByte n | n < 256 = (int16ToHexByte(n / 16)) <> (int16ToHexByte(n % 16))
+int256ToHexByte n = error $ "Int256 should be in range [0, 256): " <> show n
+
+-- | HIDE
+intToHexByteList: Int -> [BytesHex]
+intToHexByteList n | n < 256 = [int256ToHexByte n]
+intToHexByteList n = int256ToHexByte(n % 256) :: intToHexByteList(n / 256)
+
+instance HasToHex Int where
+  toHex n = Text.implode(reverse(intToHexByteList n))
```
[the remark this answers is not in the document]
Comment (carlpulley-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml:131]: Oh, I think I've figured out what you're trying to ask for here:
- `toHex 1` should be `("00" * 7) <> "01"` (as this is the encoding for a `Int64` as it has an 8 byte byte array)

That's not an issue, now I understand I can refactor to fix this.

**Comment by carlpulley-da** [comment]
Adding the warning will (due to daml-script complications) be handled on a separate PR. @samuel-williams-da is currently working to resolve that issue.

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml (hunk 11 above)
Comment (carlpulley-da) [reply sdk/compiler/damlc/tests/daml-test-files/HexString.daml:42]: Currently planning to manage this on https://github.com/digital-asset/daml/issues/20947

File: sdk/compiler/damlc/tests/daml-test-files/HexString.daml (hunk 12 above)
Comment (carlpulley-da) [reply sdk/compiler/damlc/tests/daml-test-files/HexString.daml:66]: Currently planning to manage this on https://github.com/digital-asset/daml/issues/20947

**Comment by carlpulley-da** [comment]
Currently planning to manage all unfinished tests on https://github.com/digital-asset/daml/issues/20947

File: sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml (hunk 15 above)
Comment (dylant-da) [reply sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml]: Fun fact: a slide rule came in handy on this one :)
